cirq/protocols/apply_unitary.py

```diff
@@ -85,28 +85,138 @@ def __init__(self,
                 dtype as the target tensor.
             axes: Which axes the unitary effect is being applied to (e.g. the
                 qubits that the gate is operating on).
+
         """
         self.target_tensor = target_tensor
         self.available_buffer = available_buffer
         self.axes = tuple(axes)
 
     @staticmethod
-    def default(num_qubits: int) -> 'ApplyUnitaryArgs':
-        """A default instance starting in state |0⟩."""
+    def default(num_qubits: Optional[int] = None,
+                *,
+                qid_shape: Optional[Tuple[int, ...]] = None
+               ) -> 'ApplyUnitaryArgs':
+        """A default instance starting in state |0⟩.
+
+        Specify exactly one argument.
+
+        Args:
+            num_qubits: The number of qubits to make space for in the state.
+            qid_shape: The shape of the state, specifying the dimension of each
+                qid."""
+        if (num_qubits is None) == (qid_shape is None):
+            raise TypeError('Specify exactly one of num_qubits or qid_shape.')
+        if num_qubits is not None:
+            qid_shape = (2,) * num_qubits
+        qid_shape = cast(Tuple[int, ...], qid_shape)  # Satisfy mypy
+        num_qubits = len(qid_shape)
         state = linalg.one_hot(index=(0,) * num_qubits,
-                               shape=(2,) * num_qubits,
+                               shape=qid_shape,
                                dtype=np.complex128)
         return ApplyUnitaryArgs(state, np.empty_like(state), range(num_qubits))
 
-    def subspace_index(self, little_endian_bits_int: int
-                       ) -> Tuple[Union[slice, int, 'ellipsis'], ...]:
+    def for_operation_with_qid_shape(self, indices: Iterable[int],
```

Comment:
Does it make sense for this method to be exposed to users? Will they ever call it correctly? We could make it private.

Reply:
Made private.